[WebGPU Swift] Swift version does not sub allocate buffers in WebGPU.Queue.writeBuffer

mwyrzykowski opened the pull request:
#### 03073c42585a6ddb1995d502afac66223cb19e3a
<pre>
[WebGPU Swift] Swift version does not sub allocate buffers in WebGPU.Queue.writeBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=297663">https://bugs.webkit.org/show_bug.cgi?id=297663</a>
<a href="https://rdar.apple.com/158779152">rdar://158779152</a>

Reviewed by Tadeu Zagallo.

Swift version does not sub allocate buffers in WebGPU.Queue.writeBuffer

Add support for this otherwise using WebGPU Swift will be a perf regression

* Source/WebGPU/WebGPU/Queue.h:
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::metalDevice const):
* Source/WebGPU/WebGPU/Queue.swift:
(WebGPU.writeBuffer(_:bufferOffset:data:)):

Canonical link: <a href="https://commits.webkit.org/300351@main">https://commits.webkit.org/300351@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b5013b64938e7647b34c2d8f89d90885d652cf6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119295 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29637 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125533 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71365 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0f0d60dc-afd7-4ff5-9071-37fef523458a) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39679 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47563 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90643 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59987 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f31276e0-f4a8-4fa8-bfbc-c363ca7980cf) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122247 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31652 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106952 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71066 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c5689ac7-54ed-4311-93d1-1e60cf8b6a53) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30694 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69182 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101104 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25252 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128540 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46213 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34958 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99209 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46578 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103151 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98985 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25699 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44445 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22457 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42780 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46076 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51776 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45541 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48891 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47228 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.swift:
```diff
@@ -30,33 +30,37 @@ private let largeBufferSize = 32 * 1024 * 1024
-                    data.__dataUnsafe(), data.size(), MTLResourceOptions.storageModeShared)
-        else {
+
+        let noCopy = unsafe data.size() >= largeBufferSize
```
I'm not sure how to reduce `unsafe` usages here. Even calling `.size()` I am seeing a warning that the call is unsafe :(

adetaylor replied:
I made a minimal test case for `std::span::size` being unsafe and filed rdar://159839254. We'll see what comes back!

tadeuzagallo replied:
Any operation that touches a `std::span` will be unsafe since it's an unsafe type (it doesn't carry lifetime information), so even if it's just accessing a property it doesn't matter. I think it's ok to land as is, this file still needs to be reworked, we might have need to tweak the API to reduce the unsafeness.

geoffreygaren commented:
I agree with @adetaylor that getting to the point where `unsafe` is super minimal, and treated as bad code smell, is our near term goal. However, I think we’re still figuring out how to do that in this code. (I’m working on this now.) Until we have a sufficient recommended idiom / design, I think it’s ok to proceed with `unsafe` here.

mwyrzykowski committed:
[WebGPU Swift] Swift version does not sub allocate buffers in WebGPU.Queue.writeBuffer
https://bugs.webkit.org/show_bug.cgi?id=297663
rdar://158779152

Reviewed by Tadeu Zagallo.

Swift version does not sub allocate buffers in WebGPU.Queue.writeBuffer

Add support for this otherwise using WebGPU Swift will be a perf regression

* Source/WebGPU/WebGPU/Queue.h:
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::metalDevice const):
* Source/WebGPU/WebGPU/Queue.swift:
(WebGPU.writeBuffer(_:bufferOffset:data:)):

Canonical link: https://commits.webkit.org/300351@main
